custom-targets/vertex-ai/model-deployer/vertexai.go

```diff
@@ -0,0 +1,199 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"google.golang.org/api/aiplatform/v1"
+	"google.golang.org/api/option"
+	"os"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+// deployModelFromManifest loads the file provided in `path` and returns the parsed DeployModelRequest
+// from the data.
+func deployModelFromManifest(path string) (*aiplatform.GoogleCloudAiplatformV1DeployModelRequest, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading manifest file: %v", err)
+	}
+
+	deployModelRequest := &aiplatform.GoogleCloudAiplatformV1DeployModelRequest{}
+
+	if err = yaml.Unmarshal(data, deployModelRequest); err != nil {
+		return nil, fmt.Errorf("unable to parse deploy model deployModelRequest from manifest file: %v", err)
+	}
+
+	return deployModelRequest, nil
+}
+
+// determinePreviousModel queries the provided Vertex AI endpoint to determine the model that was previously
+// deployed.
+func determinePreviousModel(service *aiplatform.Service, endpointName, currentModel string) (string, error) {
+	endpoint, err := service.Projects.Locations.Endpoints.Get(endpointName).Do()
+	if err != nil {
+		return "", fmt.Errorf("unable to fetch endpoint: %v", err)
+	}
+
+	deployedModels := map[string]*aiplatform.GoogleCloudAiplatformV1DeployedModel{}
+
+	for _, dm := range endpoint.DeployedModels {
+		modelNameWithVersion := resolveDeployedModelNameWithVersion(dm)
+		deployedModels[modelNameWithVersion] = dm
+	}
+
+	delete(deployedModels, currentModel)
+
+	if len(deployedModels) != 1 {
+		return "", fmt.Errorf("unable to resolve previous deployed currentModel to canary against. Not including the current currentModel to be deployed, the endpoint has %d deployed models but expected only one", len(deployedModels))
+	}
+
+	firstModel := []*aiplatform.GoogleCloudAiplatformV1DeployedModel{}
+
+	for _, dm := range deployedModels {
+		firstModel = append(firstModel, dm)
+	}
+	return firstModel[0].Id, nil
+}
+
+// resolveDeployedModelNameWithVersion returns the model resource name associated with the  provided DeployedModel
+// with its version ID attached.
+func resolveDeployedModelNameWithVersion(deployedModel *aiplatform.GoogleCloudAiplatformV1DeployedModel) string {
+	if strings.Contains(deployedModel.Model, "@") {
+		return deployedModel.Model
+	}
+	return fmt.Sprintf("%s@%s", deployedModel.Model, deployedModel.ModelVersionId)
+}
+
+// resolveModelWithVersion returns the model resource name its version ID attached.
+func resolveModelWithVersion(model *aiplatform.GoogleCloudAiplatformV1Model) string {
+	if strings.Contains(model.Name, "@") {
+		return model.Name
+	}
+	return fmt.Sprintf("%s@%s", model.Name, model.VersionId)
+}
+
+// extracts the region from the model region name.
+func fetchRegionFromModel(modelName string) (string, error) {
```

Comment:
nit: for the functions that don't make any API calls remove "fetch" since it reads as if it's fetching it from some service.